chains/sequential.go

```diff
@@ -0,0 +1,205 @@
+// where the output of one chain is the input of the next.
+type SequentialChain struct {
+	chains     []Chain
+	inputKeys  []string
+	outputKeys []string
+	memory     schema.Memory
+}
+
+func NewSequentialChain(chains []Chain, inputKeys []string, outputKeys []string) (*SequentialChain, error) {
+	if err := validateSeqChain(chains, inputKeys, outputKeys); err != nil {
+		return nil, err
+	}
+
+	return &SequentialChain{
+		chains:     chains,
+		inputKeys:  inputKeys,
+		outputKeys: outputKeys,
+		memory:     memory.NewSimple(),
+	}, nil
+}
+
+func validateSeqChain(chain []Chain, inputKeys []string, outputKeys []string) error {
+	knownKeys := toSet(inputKeys)
+
+	for i, c := range chain {
+		// Check that chain has input keys that are in knownKeys
+		missingKeys := difference(c.GetInputKeys(), knownKeys)
+		if len(missingKeys) > 0 {
+			return fmt.Errorf(
+				"%w: chain at index %d is missing required input keys: %v",
+				ErrChainInitialization, i, missingKeys,
+			)
+		}
+
+		// Check that chain does not have output keys that are already in knownKeys
+		overlappingKeys := intersection(c.GetOutputKeys(), knownKeys)
+		if len(overlappingKeys) > 0 {
+			return fmt.Errorf(
+				"%w: chain at index %d has output keys that already exist: %v",
+				ErrChainInitialization, i, overlappingKeys,
+			)
+		}
+
+		// Add the chain's output keys to knownKeys
+		for _, key := range c.GetOutputKeys() {
+			knownKeys[key] = struct{}{}
+		}
+	}
+
+	// Check that outputKeys are in knownKeys
+	for _, key := range outputKeys {
+		if _, ok := knownKeys[key]; !ok {
+			return fmt.Errorf("%w: output key %s is not in the known keys", ErrChainInitialization, key)
+		}
+	}
+
+	return nil
+}
+
+// toSet converts a list to a set.
+func toSet(list []string) map[string]struct{} {
+	set := make(map[string]struct{}, 0)
+	for _, v := range list {
+		set[v] = struct{}{}
+	}
+	return set
+}
+
+// difference returns the elements in list that are not in set.
+func difference(list []string, set map[string]struct{}) []string {
+	diff := make([]string, 0)
+	for _, v := range list {
+		if _, ok := set[v]; !ok {
+			diff = append(diff, v)
+		}
+	}
+	return diff
+}
+
+// intersection returns the elements in list that are in set.
+func intersection(list []string, set map[string]struct{}) []string {
+	intersection := make([]string, 0)
+	for _, v := range list {
+		if _, ok := set[v]; ok {
+			intersection = append(intersection, v)
+		}
+	}
+	return intersection
+}
+
+// Call runs the logic of the chains and returns the outputs. This method should
+// not be called directly. Use rather the Call, Run or Predict functions that
+// handles the memory and other aspects of the chain.
+func (c *SequentialChain) Call(ctx context.Context, inputs map[string]any, options ...ChainCallOption) (map[string]any, error) { //nolint:lll
+	for _, chain := range c.chains {
+		var err error
+		inputs, err = Call(ctx, chain, inputs, options...)
```

Comment:
Tbh I agree this could be clearer. I (blindly) did this to keep some level of parity with Python [source](https://github.com/hwchase17/langchain/blob/9cb2347453015d661520ec78d0f2f1e5968e0d92/langchain/chains/sequential.py#L176-L177):
```python
_input = chain.run(_input, callbacks=_run_manager.get_child(f"step_{i+1}"))
```

Is this clearer?
```go
var outputs map[string]any
var err error
for _, chain := range c.chains {
  outputs, err = Call(ctx, chain, inputs, options...)
  if err != nil {
    return nil, err
  }
  inputs = outputs
}
```

Reply:
@tobiade adding a comment on the original might be good enough